lib/nmatrix/nmatrix.rb

```diff
@@ -90,6 +90,62 @@ def size(shape)
       shape = [shape,shape] unless shape.is_a?(Array)
       (0...shape.size).inject(1) { |x,i| x * shape[i] }
     end
+
+    # Make N-D coordinate arrays for vectorized evaluations of
+    # N-D scalar/vector fields over N-D grids, given N one-dimensional
+    # coordinate arrays arrs. N > 1.
+    #
+    # call-seq:
+    #     meshgrid(arrs) -> Array of NMatrix
+    #     meshgrid(arrs, options) -> Array of NMatrix
+    #
+    # * *Arguments* :
+    #   - +arrs+ -> Array of N one-dimensional coordinate arrays (Array or NMatrix)
+    #   - +options+ -> Hash with options (:sparse and :ij both Boolean, false by default)
+    # * *Returns* :
+    #   - Array of N N-D NMatrixes
+    # * *Examples* :
+    #     x, y = NMatrix::meshgrid([[1, 2, 3], [4, 5]])
+    #     x.to_a #<= [[1, 2, 3], [1, 2, 3]]
+    #     y.to_a #<= [[4, 4, 4], [5, 5, 5]]
+    #
+    # * *Using* *options* :
+    #
+    #     x, y = NMatrix::meshgrid([[1, 2, 3], [4, 5]], sparse: true)
+    #     x.to_a #<= [[1, 2, 3]]
+    #     y.to_a #<= [[4], [5]]
+    #
+    #     x, y = NMatrix::meshgrid([[1, 2, 3], [4, 5]], ij: true)
+    #     x.to_a #<= [[1, 1], [2, 2], [3, 3]]
+    #     y.to_a #<= [[4, 5], [4, 5], [4, 5]]
+    def meshgrid(arrs, options = {})
+      raise(ArgumentError, 'Expected at least 2 arrays.') if arrs.size < 2
+      mats = arrs.map do |arr|
+        case arr
+        when NMatrix
+          arr
+          raise(ArgumentError, 'Expected one-dimensional arrays.') if arr.effective_dim != 1
+        when Array
+          NMatrix.new([arr.size], arr)
+        else
+          raise(ArgumentError, 'Expected one-dimensional arrays.')
+        end
+      end
+      mats[0], mats[1] = mats[1], mats[0] unless options[:ij]
+      new_dim = mats.size
+      lengths = mats.map(&:size)
+      result = mats.map.with_index do |matrix, axis|
+        new_shape = Array.new(new_dim, 1)
+        new_shape[axis] = lengths[axis]
+        new_arr = matrix.reshape(new_shape)
+        new_dim.times do |i|
+          new_arr = new_arr.repeat(lengths[i], i) if axis != i
+        end unless options[:sparse]
```

Comment:
IMO programming and especially Ruby is all about using existing code as frequent as possible and using Arrays in this case I will reinvent the wheel. NMatrix already invented one so I don't think #meshgrid should be done with Arrays.
I'm totally agree with you in aim to minimize memory reallocation. Maybe it can be achieved some other way? I can make #repeat to take more arguments (optional, #repeat(Fixnum, Fixnum) will continue working) in order to replace chains like this (2x NMatrix.new calls)
> NMatrix([1,1],[3]).repeat(3, 0).repeat(3, 1)

with something like that (1x NMatrix.new call)
> NMatrix([1, 1], [3]).repeat(3, 0, 3, 1)

or
> NMatrix([1, 1], [3]).repeat({count: 3, axis: 0}, {count: 3, axis: 1})